### DISCUSSION TITLE
T:Slim X2 Integration

### DISCUSSION THREAD

**Discussion by Miquest** [NightscoutFoundation/xDrip#2154] [Ideas]
Dear developers, 
I will get my new insulin pump in 2 days!  The T:Slim X2.  I would like a function with which the T:Slim can be read out directly with the cell phone and I can see the basal rates and insulin delivery directly.  Unfortunately I'm not a developer, but maybe you know a solution?

Kind regards,
Thorin

**Comment by Navid200** [comment]
I believe this is related if not exactly the same:
https://github.com/NightscoutFoundation/xDrip/issues/904

**Reply by Miquest** [reply]
Okay, thank you...
I see the problem with the API, which doesn't exist...
But maybe it's helpful to implement Glokoo?
They have a open API, I think...

**Reply by Navid200** [reply]
@jwoglom

**Reply by jwoglom** [reply]
I'd recommend taking a look at https://github.com/jwoglom/tconnectsync as it is the closest thing to a direct t:slim X2/xDrip integration we have today.

**Reply by Miquest** [reply]
I'm living in Europe, so unfortunately I'm not able to use the t:connect mobile application...

**Reply by Navid200** [reply]
Why is being in Europe a limitation?  Would you please provide a reference to the limitation for others who may wonder?

**Reply by tolot27** [reply]
The app is only certified in the US. It is indeed possible to create an account at https://tconnect.tandemdiabetes.com from outside of the US but it is impossible to connect a pump to this account. They maintain a list of serial numbers and countries they have sold the pumps and block all from outside the US. I've asked them how to get data in real-time but they just directed me to my local reseller (VitalAire). VitalAire suggested to use Diasend but they do not know that you cannot get data from the pump with Diasend running on the phone, only if you have a computer.
Hence, outside from the US you cannot get data from the pump in real-time.

**Reply by Miquest** [reply]
Okay, thank you for the quick replys...
So, I have to wait for the Update in Europe.

**Reply by jonprogers** [reply]
I'm in the UK with a tSlim pump. We only have access to the gluko (formerly diasend) platform to access pump data. This is done by occasional wired pc download so we have no real time information available from the pump. Having this would make a good pump even more valuable in the fight against errant glucose values, the people deciding on these things obviously don't live with T1D!

**Reply by BenoitAnastay** [reply]
I'm wondering if there is a way of patching the t:connect app with a personalized API endpoint, if each pump have RSA key pair to secure data exchange it might be hard to hack, if there is no encryption we only need to have an US pump to reverse engine the API and create our own.

It's sad they didn't get a EU located server as Dexcom did.

I'll relay on glooko to extract the pump history but what we need most is predicting capacities.

The unavailability of the t:connect app and the disallowed Control-IQ make this pump not better than an omnipod afterall here in France

**Reply by jwoglom** [reply]
@BenoitAnastay if you're at all interested in taking this on, take a look at https://github.com/jwoglom/tconnectpatcher which can be used to modify parts of the tconnect Android app version 1.2. It strips the TLS cert validation, so you can either find-replace the .tandemdiabetes.com URLs with your own or run a local DNS server to point them somewhere else, if you're interested in trying to build a reimplementation of their server to accept pump events from scratch.

We can also get read-only data from the pump now over Bluetooth: https://github.com/jwoglom/pumpx2. That is another pathway to extracting pump data which doesn't involve the tconnect app at all.

**Reply by BenoitAnastay** [reply]
@jwoglom  Nice work for the Bluetooth protocol, we can use intent to send datas to xDrip+ or make an nightscoot integration.

For the t:connect RE I would really need to observe "normal" communication between the app and the server and it's kinda hard to implement all required functionalities, nightscoot and xDrip are still a better pathway for monitoring purpose.

If the pump is able to receive bolus orders it can be integrated withing Android APS

**Reply by jwoglom** [reply]
@BenoitAnastay we don't have the mobile bolus functionality reverse engineered yet, but it should be doable. I'm already working to try and complete the first steps necessary to integrate the tslim x2 with AAPS, but prioritizing finishing up the BT reverse engineering work first.

There is also a bit of work to do in fully solidifying the parsing which is done to read history log events, like past pump basal/bolus events and etc. This is all being tracked in the pumpx2 repo.

The pumpx2 code is being intentionally built as a library that can be consumed by other projects. I could imagine myself or someone else building an app a la what xDrip is to the official Dexcom app, which uses the library to connect with the pump and upload data to xDrip or directly to Nightscout.

**Reply by jwoglom** [reply]
Also would encourage anyone in this thread interested to join #tslim-x2-dev in the Nightscout/WeAreNotWaiting discord, since this isn't directly related to xDrip and is somewhat off-topic for this forum: https://discord.gg/KyZGdhyt44

**Reply by Miquest** [reply]
I would be happy if there would be a way to get data directly from the Insulin Pump.
So I could maybe workaround my Bluetooth Connection Problems with the G6.
Great work with the Bluetooth Data Reciver, @jwoglom ! Thank you so much!

@jamorham, in the Nightly Versions is a "Companion" Interface to recive data.
Could you configure it that it can get Data from @jwoglom's Data reading service?
I would be very happy if you do that...
With bolusing Diagrams and Blood Glucose reading...

**Reply by jwoglom** [reply]
Don't want to put anything on the xDrip devs yet because this isn't yet ready for primetime. But don't worry, I daily drive xDrip and have contributed a bit to xDrip in the past, and fully want to get this all integrated together.

**Reply by Miquest** [reply]
Or maybe with a separate menu to recive bolusing data and the current status?
I just read that there is no Blood Glucose interface from the pump

**Reply by Miquest** [reply]
> Don't want to put anything on the xDrip devs yet because this isn't yet ready for primetime. But don't worry, I daily drive xDrip and have contributed a bit to xDrip in the past, and fully want to get this all integrated together.


Okay, good luck and thank you so much!
☺️
